libtenzir/builtins/aggregation-functions/approximate_median.cpp

```diff
@@ -0,0 +1,114 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2024 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/aggregation_function.hpp>
+#include <tenzir/plugin.hpp>
+
+#include <arrow/util/tdigest.h>
+
+namespace tenzir::plugins::approximate_median {
+
+namespace {
+
+template <basic_type Type>
+class approximate_median_function final : public aggregation_function {
+public:
+  explicit approximate_median_function(type input_type) noexcept
+    : aggregation_function(std::move(input_type)), tdigest_{} {
+    // nop
+  }
+
+private:
+  auto output_type() const -> type override {
+    return input_type();
+  }
+
+  auto add(const data_view& view) -> void override {
+    using view_type = tenzir::view<type_to_data_t<Type>>;
+    if (caf::holds_alternative<caf::none_t>(view)) {
+      return;
+    }
+    const auto x = static_cast<double>(caf::get<view_type>(view));
+    if constexpr (std::is_same_v<Type, double_type>) {
+      if (std::isnan(x)) {
+        return;
+      }
+    }
+    tdigest_.Add(x);
+  }
+
+  auto add(const arrow::Array& array) -> void override {
+    const auto& typed_array = caf::get<type_to_arrow_array_t<Type>>(array);
+    for (auto&& value : values(Type{}, typed_array)) {
+      if (not value) {
+        continue;
+      }
+      if constexpr (std::is_same_v<Type, double_type>) {
+        if (std::isnan(*value)) {
+          continue;
+        }
+      }
+      tdigest_.Add(static_cast<double>(*value));
+    }
+  }
+
+  auto finish() && -> caf::expected<data> override {
+    if (tdigest_.is_empty()) {
+      return data{};
+    }
+    return data{static_cast<type_to_data_t<Type>>(tdigest_.Quantile(0.5))};
+  }
+
+  arrow::internal::TDigest tdigest_;
+};
+
+class plugin : public virtual aggregation_function_plugin {
+  auto name() const -> std::string override {
+    return "approximate_median";
```

Comment:
```suggestion
    return "approx_median";
```

Anything longer than that is going to be painful to write.

Reply:
I've chosen to use the same names used by Pandas, Polars, and Arrow. I'm reluctant to change this.

Reply:
Good reason.